xds/src/main/java/io/grpc/xds/XdsLbState.java

```diff
@@ -129,50 +125,119 @@ final XdsComms shutdownAndReleaseXdsComms() {
+      Locality locality = (Locality) o;
+      return Objects.equal(region, locality.region)
+          && Objects.equal(zone, locality.zone)
+          && Objects.equal(subzone, locality.subzone);
     }
 
-
     @Override
-    public boolean hasSubchannel(Subchannel subchannel) {
-      // TODO: impl
-      return false;
+    public int hashCode() {
+      return Objects.hashCode(region, zone, subzone);
     }
 
     @Override
-    public void shutdown() {
-      // TODO: impl
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("region", region)
+          .add("zone", zone)
+          .add("subzone", subzone)
+          .toString();
     }
   }
 
   /**
-   * The interface of {@link XdsLbState.SubchannelStoreImpl} that is convenient for testing.
+   * Information about the locality from EDS response.
    */
-  public interface SubchannelStore {
+  static final class LocalityInfo {
+    final List<EquivalentAddressGroup> eags;
+    final List<Integer> endPointWeights;
+    final int localityWeight;
+
+    LocalityInfo(Collection<LbEndpoint> lbEndPoints, int localityWeight) {
+      List<EquivalentAddressGroup> eags = new ArrayList<>(lbEndPoints.size());
+      List<Integer> endPointWeights = new ArrayList<>(lbEndPoints.size());
+      for (LbEndpoint lbEndPoint : lbEndPoints) {
+        eags.add(lbEndPoint.eag);
+        endPointWeights.add(lbEndPoint.endPointWeight);
+      }
+      this.eags = Collections.unmodifiableList(eags);
+      this.endPointWeights = Collections.unmodifiableList(new ArrayList<>(endPointWeights));
+      this.localityWeight = localityWeight;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      LocalityInfo that = (LocalityInfo) o;
+      return localityWeight == that.localityWeight
+          && Objects.equal(eags, that.eags)
+          && Objects.equal(endPointWeights, that.endPointWeights);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(eags, endPointWeights, localityWeight);
+    }
+  }
 
-    boolean hasReadyBackends();
+  static final class LbEndpoint {
+    final EquivalentAddressGroup eag;
+    final int endPointWeight;
+    final HealthStatus healthStatus;
 
-    boolean hasNonDropBackends();
+    LbEndpoint(io.envoyproxy.envoy.api.v2.endpoint.LbEndpoint lbEndpointProto) {
 
-    boolean hasSubchannel(Subchannel subchannel);
+      this(
+          new EquivalentAddressGroup(ImmutableList.of(fromEnvoyProtoAddress(lbEndpointProto))),
+          lbEndpointProto.getLoadBalancingWeight().getValue(),
+          lbEndpointProto.getHealthStatus());
+    }
+
+    @VisibleForTesting
+    LbEndpoint(EquivalentAddressGroup eag, int endPointWeight, HealthStatus healthStatus) {
+      this.eag = eag;
+      this.endPointWeight = endPointWeight;
+      this.healthStatus = healthStatus;
```

Comment:
If this field is not used at all, we should remove it.

Reply:
Done.